File: assignment_1/merge.py
# 리스트를 반으로 나누고 앞 뒤 따로따로 정렬해서 재귀로 합치는 방식
# time : O(n * log n) : log_2 n 번 재귀 호출(n개의 요소 리스트를 2개로 나누니까), 각 재귀 호출마다 n개를 병합해야하니까 
# space : O(n) : 새 리스트 생성해서 결과물 넣음  

# merge_sort function 제작
def merge_sort(list):
    # 만약 요소가 작으면 걍 반환
    if (len(list) <= 1) :
        return list
    
    # 중앙을 pivot으로 잡고 left랑 right로 나누기 
    pivot = len(list) // 2
    left_list = merge_sort(list[:pivot])
    right_list = merge_sort(list[pivot:])

    # 만약 정렬을 해야한다면, 새 리스트 하나 만들고
    result = []
    # i 는 left_list 의 index, j는 right_list 의 index
    i = j = 0

    # 양쪽 list의 맨 앞끼리 비교해서 새 리스트에 집어넣기
    while ((i < len(left_list)) and (j < len(right_list))):
        if (left_list[i] <= right_list[j]):
            result.append(left_list[i])
            i += 1
        else:
            result.append(right_list[j])
            j += 1
        
    # 한 list 가 다 추가된 다음에는 남은거 마지막에 싹 추가 
    result.extend(left_list[i:])
    result.extend(left_list[j:])

    return result

File: assignment_1/merge.cpp
#include <vector>
using namespace std;

// merge_sort function 제작
vector<int> merge_sort(vector<int> list) {
    // 만약 요소가 작으면 걍 반환
    if (list.size() <= 1) {
        return list;
    }

    // 중앙을 pivot으로 잡고 left랑 right로 나누기 
    int pivot = list.size() / 2;
    vector<int> left_list(list.begin(), list.begin() + pivot);
    vector<int> right_list(list.begin() + pivot, list.end());

    left_list = merge_sort(left_list);
    right_list = merge_sort(right_list);

    // 만약 정렬을 해야한다면, 새 리스트 하나 만들고
    vector<int> result;
    // i 는 left_list 의 index, j는 right_list 의 index
    int i = 0, j = 0;

    // 양쪽 list의 맨 앞끼리 비교해서 새 리스트에 집어넣기
    while ((i < left_list.size()) && (j < right_list.size())) {
        if (left_list[i] <= right_list[j]) {
            result.push_back(left_list[i]);
            i += 1;
        } else {
            result.push_back(right_list[j]);
            j += 1;
        }
    }

    // 한 list 가 다 추가된 다음에는 남은거 마지막에 싹 추가 
    result.insert(result.end(), left_list.begin() + i, left_list.end());
    result.insert(result.end(), right_list.begin() + j, right_list.end());

    return result;
}
